content/integrations/integrating-npm-with-external-services/docker-and-private-modules.mdx

```diff
@@ -4,87 +4,71 @@ redirect_from:
 
-```
+```docker
 ENV NPM_TOKEN=00000000-0000-0000-0000-000000000000
 ```
 
 However, this doesn't work as you would expect, because you want the npm install to occur when you run `docker build`, and in this instance, `ENV` variables aren't used, they are set for runtime only.
 
-Instead of run-time variables, you must use a different way of passing environment variables to Docker, available since Docker 1.9: the [ARG parameter](https://docs.docker.com/engine/reference/builder/#arg).
-
-## Create and check in a project-specific .npmrc file
-
-A complete example that will allow you to use `--build-arg` to pass in your NPM_TOKEN requires adding a `.npmrc` file to the project.
+Instead of run-time variables, you must use Docker build secrets.
 
-Use a project-specific `.npmrc` file with a variable for your token to securely authenticate your Docker image with npm.
-
-1. In the root directory of your project, create a custom <a href="https://docs.npmjs.com/cli-documentation/files/npmrc">`.npmrc`</a> file with the following contents:
+## Update the Dockerfile
 
-   ```
-   //registry.npmjs.org/:_authToken=${NPM_TOKEN}
-   ```
+The Dockerfile that takes advantage of this has a few more lines in it than the earlier example that allows us to use your global `.npmrc` and the access token created when running `npm login` command (if you haven't run it already - do so before moving on).
 
-   **Note:** that you are specifying a literal value of `${NPM_TOKEN}`.  The npm cli will replace this value with the contents of the `NPM_TOKEN` environment variable.  Do **not** put a token in this file.
+```dockerfile
+# https://docs.npmjs.com/docker-and-private-modules
+FROM node:14
 
-2. Check in the `.npmrc` file.
+ENV APP_HOME="/app/"
 
-## Update the Dockerfile
+WORKDIR ${APP_HOME}
 
-The Dockerfile that takes advantage of this has a few more lines in it than the earlier example that allows us to use the `.npmrc` file and the `ARG` parameter:
+COPY package*.json ${APP_HOME}/
 
-```
-FROM node
+RUN --mount=type=secret,id=npm,target=/root/.npmrc npm install
 
-ARG NPM_TOKEN  
-COPY .npmrc .npmrc  
-COPY package.json package.json  
-RUN npm install  
-RUN rm -f .npmrc
+COPY . ${APP_HOME}/
 
-# Add your source files
-COPY . .  
 CMD npm start
+
 ```
 
-This adds the expected `ARG NPM_TOKEN`, but also copies the `.npmrc` file, and removes it when `npm install` completes.
+This will configure your Dockerfile to receive `.npmrc` file via build secrets, that will leave no trace after npm dependency installation is done.
 
 ## Build the Docker image
 
 To build the image using the above Dockerfile and the npm authentication token, you can run the following command. Note the `.` at the end to give `docker build` the current directory as an argument.
 
-```
-docker build --build-arg NPM_TOKEN=${NPM_TOKEN} .
+```shell
+docker build . -t secure-app-secrets:1.0 --secret id=npm,src=$HOME/.npmrc
```

Comment:
should we be encouraging folks to be using their $HOME/.npmrc? I'd imagine they might be managing secrets specific for this deployment 🤔. In particular they might want to use a read-only token or automation token rather than the "full access" token that is used by default when logged in. This could also cause additional complications if 2FA is enabled on the account.